File: src/life.scala
type State[A] = A
type Position = List[Int]
type Cell[A] = (State[A], Position)
type Rule[A] = (State[A], List[Cell[A]]) => (State[A])
type World[A] = List[Cell[A]]
type Neighbourhood[A] = (World[A], Position) => (List[Cell[A]])

def apply_rule[A] (world: World[A], rule: Rule[A], neighbourhood:Neighbourhood[A]) (cell:Cell[A]) = {
  val (state, position) = cell
  val neighbours = neighbourhood (world, position)
  val newState = rule (state, neighbours)
  (newState, position)
}

def step_time[A] (world:World[A], rule:Rule[A], neighbourhood:Neighbourhood[A]) =
  world.map(apply_rule (world, rule, neighbourhood))

//--------------

abstract class LifeStates
case object Alive extends LifeStates
case object Dead extends LifeStates

def rule_life (state: State[LifeStates], neighbours: List[Cell[LifeStates]]) = {
  def isAlive (cell: Cell[LifeStates]) = cell match {
    case (Alive, _) => true
    case _ => false
  }

  val count = 
		neighbours.foldLeft(0) { (a,c) => a + (if (isAlive (c)) 1 else 0) }

  state match {
		case Alive if count < 2 || count > 3 => Dead
    case Dead if count == 3 => Alive
    case other => other
  }
}

def neighbourhood_life (world: World[LifeStates], pos: Position) = {
  def is_near (c1: Int, c2: Int): Boolean = math.abs(c1 - c2) <= 1
  world.filter { case (_,p) =>
      (pos.zip(p).foldLeft(true) { (r, el) => {
          val (c1,c2) = el
          r && (is_near(c1,c2)) 
      	}
      }) && (p != pos)
  }
}

//-------------

def generate_world (str: String) = {
	def generate_state (ch: Char) = ch match {
  	case '#' => Alive
    case _ => Dead
  }
  
  str.split('\n').zipWithIndex.flatMap { case (line, i) =>
  	line.zipWithIndex.map { case (c, j) => (generate_state(c), List(i, j)) }
  }.toList
}

val initial_state = generate_world ("""
0#00000000
00#0000000
###0000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000""")

def print_world (world: World[LifeStates]) = {
  def print_cell (state: State[LifeStates]) = state match {
		case Alive => '#'
    case Dead => '0'
  }
  
  val len = world.map { case (_, p) => p.last }.max
  val res = world.foldLeft("") { case (acc, (s, p)) => acc + (print_cell (s)) + (if (p.last == len) '\n' else "") }
	println (res)
}

def main () =
	print_world (step_time (initial_state, rule_life, neighbourhood_life))

main ()

File: src/life.fs
type State<'a> = State of 'a
type Position = int list
type Cell<'a> = State<'a> * Position
type Rule<'a> = State<'a> -> Cell<'a> list -> State<'a>
type World<'a> = World of Cell<'a> list
type Neighbourhood<'a> = World<'a> -> Position -> Cell<'a> list


let apply_rule world rule neighbourhood (state, position) =
  let neighbours = neighbourhood world position
  let state' = rule state neighbours
  (state', position)

let step_time (World world) rule neighbourhood =
  List.map (apply_rule (World world) rule neighbourhood) world 
  |> World

//--------------

type LifeStates = 
  | Alive
  | Dead

let rule_life (State state) neighbours =
  let isAlive ((State state), _) = 
    match state with
    | Alive -> true
    | _ -> false 

  let count =
    List.fold (fun a c -> a + if (c |> isAlive) then 1 else 0) 0 neighbours
    
  match state with
  | Alive when count < 2 || count > 3 -> Dead
  | Dead when count = 3 -> Alive 
  | state -> state
  |> State

let neighbourhood_life (World world) pos =
  let is_near c1 c2 = abs (c1-c2) <= 1

  world
  |> List.filter (fun (_,p) ->
      List.fold2 (fun r c1 c2 -> r && (is_near c1 c2)) true p pos && p <> pos
  )

//----------

let generate_world (str: System.String) =
  let generate_state = function
    | '#' -> Alive |> State
    | _ -> Dead |> State

  str.Split '\n'
  |> Seq.mapi (fun i line -> 
    line
    |> Seq.mapi (fun j c -> ((generate_state c), [i;j]))
  )
  |> Seq.collect id
  |> Seq.toList
  |> World

let initial_state = generate_world @"
0#00000000
00#0000000
###0000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000"  

let print_world (World world) = 
  let print_cell = function
    | Alive -> '#'
    | _ -> '0'

  let last ls = ls |> List.rev |> List.head
  let len = world |> List.map (fun (_,p) -> last p) |> List.max

  world
  |> List.fold (fun acc ((State s), p) -> 
    sprintf "%s%c%s" acc (print_cell s) (if (last p) = len then "\n" else "")
  ) ""
  |> printfn "%s"

let main =
  step_time initial_state rule_life neighbourhood_life
  |> print_world

main